elastic/beats
# Add support for ephemeral containers in elastic agent dynamic provider

## What does this PR do?

This PR adds support for ephemeral containers in elastic agent dynamic kubernetes provider.

## Why is it important?

Ephemeral containers that are part of a pod can be monitored and be part of the published event.

## Checklist


## How to test this PR locally

1. Create a kind cluster with `"feature-gates": "EphemeralContainers=true"`. example kind config (
[kind-ephemeral.yaml.zip](https://github.com/elastic/beats/files/7105261/kind-ephemeral.yaml.zip))
2. Create en example pod 
```
cat <<EOF | kubectl create -f - 
apiVersion: v1
kind: Pod
metadata:
  name: example-pod
spec:
  containers:
  - image: mhausenblas/simpleservice:0.5.0
    name: app
EOF
```
3. Attach an ephemeral container to it `kubectl debug -it example-pod --image=busybox --target=app`
4. Run agent with pod resource enabled 
```
providers.kubernetes:
  scope: cluster
  cleanup_timeout: 360s
  resources:
    pod:
       enabled: true
```
5. Run inspect command to check what is the compiled configuration (use proper path for the elastic-agent.yml config file :
`elastic-agent  -c /etc/agent.yml -e inspect output -o default `
6. Verify in the output that mapping is produced for the ephemeral container 
```
processors:
    - add_fields:
        fields:
          container:
            id: a8d7f67674c848c27e71b7a97f3f82d9492080308a72153458a15df7485904bf
            image: busybox
            name: debugger-pbjkz
            runtime: containerd
          namespace: default
          pod:
            ip: 10.244.0.5
            labels: {}
            name: example-pod
            uid: d7a97d03-3baa-457c-b9e2-a8e80b831c23
        target: kubernetes
```

## Related issues
-  Closes https://github.com/elastic/beats/issues/27020
